File: light-control/light_strip.py
import os
import colorsys
from typing import List
from algorithms._meta.pixel import Pixel

is_dev_mode = os.getenv("MODE") == "DEVELOPMENT"

if not is_dev_mode:
    import neopixel
    import board

class DemoPixels:
    def __getitem__(self, key):
        return self._pixels[key]
    
    def __setitem__(self, key, value):
        if len(value) == 3:
            r, g, b = value
            self._pixels[key] = (int(r), int(g), int(b))
        if len(value) == 4:
            r, g, b, w = value
            self._pixels[key] = (int(r), int(g), int(b), int(w))

    def __init__(self, num_pixels, bpp):
        self._pixels = []
        self.n = num_pixels
        self.bpp = bpp
        for i in range(num_pixels):
            if (bpp == 3):
                self._pixels.append((0, 0, 0))
            elif (bpp == 4):
                self._pixels.append((0, 0, 0, 0))

class LightStrip:
    def __init__(self, strip_index: int, config):
        self.strip_index = strip_index
        self.index_start = config["index_start"]
        self.index_end = config["index_end"]

        if self.index_start < 0 or self.index_end < 0 or self.index_start == self.index_end:
            raise Exception('No light strip index may be below zero and they must be different')

        self.gpio_pin = config["gpio_pin"]
        self.bpp = config["bpp"]
        self.order = config["order"]
        self.skip = config["skip"]

        self.__generate_pixels()

        for ind in self.skip:
            self.pixels[ind] = (0, 0, 0)

    def num_pixels(self):
        return abs(self.index_end - self.index_start) + 1

    def board_pin(self):
        if is_dev_mode:
            return self.gpio_pin
        else:
            return getattr(board, self.gpio_pin)
    
    def pixel_order(self):
        if is_dev_mode:
            return self.order
        else:
            return getattr(neopixel, self.order)
    
    def __generate_pixels(self):
        if is_dev_mode:
            self.pixels = DemoPixels(self.num_pixels(), self.bpp)
        else:
            self.pixels = neopixel.NeoPixel(
                self.board_pin(),
                self.num_pixels(),
                bpp=self.bpp,
                pixel_order=self.pixel_order(),
                auto_write=False
            )
    
    def get_ranges(self):
        reversed = self.index_end < self.index_start
        lower = self.index_start if not reversed else self.index_end
        higher = self.index_end if not reversed else self.index_start
        step = 1 if not reversed else -1

        hardware_range = range(0, higher - lower + 1)
        pixel_range = range(self.index_start, self.index_end + step, step)
        return zip(hardware_range, pixel_range)
    
    def apply_lights(self, lights: List[Pixel], brightness: float):
        index_ranges = self.get_ranges()
        for (light_ind, pixel_ind) in index_ranges:
            if pixel_ind in self.skip:
                continue
            pixel = lights[pixel_ind]
            nc = colorsys.hsv_to_rgb(pixel.hue, pixel.sat, brightness * pixel.val)
            if self.bpp == 3:
                self.pixels[light_ind] = (nc[0] * 255, nc[1] * 255, nc[2] * 255)
            elif self.bpp == 4:
                white = pixel.white * brightness * 255
                self.pixels[light_ind] = (nc[0] * 255, nc[1] * 255, nc[2] * 255, white)
        
        if not is_dev_mode:
            self.pixels.show()

File: light-control-ts/src/lightStrip.ts
import { Pixel, hsvToRgb } from './algorithms/_meta/index.js'
import { LightStripConfig } from './lightConfig.js'

const isDevMode = process.env['MODE'] === 'DEVELOPMENT'

// Lazy-import piixel so the module can load in dev mode without hardware
let ws281x: import('piixel').Ws281xAPI | null = null
let StripType: typeof import('piixel').StripType | null = null

async function getPiixel() {
  if (!ws281x) {
    const piixel = await import('piixel')
    ws281x = piixel.ws281x
    StripType = piixel.StripType
  }
  return { ws281x, StripType: StripType! }
}

function packRgb(r: number, g: number, b: number): number {
  return ((r & 0xff) << 16) | ((g & 0xff) << 8) | (b & 0xff)
}

function packRgbw(r: number, g: number, b: number, w: number): number {
  return ((w & 0xff) << 24) | ((r & 0xff) << 16) | ((g & 0xff) << 8) | (b & 0xff)
}

// Terminal color preview for dev mode
function termColor(r: number, g: number, b: number): string {
  return `\x1b[48;2;${r};${g};${b}m  \x1b[0m`
}

export class LightStrip {
  readonly stripIndex: number
  readonly indexStart: number
  readonly indexEnd: number
  private readonly gpioPin: number
  private readonly bpp: 3 | 4
  private readonly order: string
  private readonly skip: Set<number>
  private pixelArray: Uint32Array
  private initialized = false

  constructor(stripIndex: number, config: LightStripConfig) {
    this.stripIndex = stripIndex
    this.indexStart = config.index_start
    this.indexEnd = config.index_end

    if (
      this.indexStart < 0 ||
      this.indexEnd < 0 ||
      this.indexStart === this.indexEnd
    ) {
      throw new Error('No light strip index may be below zero and they must be different')
    }

    this.gpioPin = config.gpio_pin
    this.bpp = config.bpp
    this.order = config.order
    this.skip = new Set(config.skip)
    this.pixelArray = new Uint32Array(this.numPixels())
  }

  numPixels(): number {
    return Math.abs(this.indexEnd - this.indexStart) + 1
  }

  private getRanges(): [number, number][] {
    const reversed = this.indexEnd < this.indexStart
    const step = reversed ? -1 : 1
    const ranges: [number, number][] = []

    let hwIdx = 0
    for (let pixelIdx = this.indexStart; pixelIdx !== this.indexEnd + step; pixelIdx += step) {
      ranges.push([hwIdx, pixelIdx])
      hwIdx++
    }
    return ranges
  }

  async initialize(): Promise<void> {
    if (this.initialized || isDevMode) {
      this.initialized = true
      return
    }
    const { ws281x, StripType } = await getPiixel()
    const stripType = (StripType as Record<string, number>)[this.order] ?? StripType.WS2811_STRIP_GRB
    ws281x.configure({
      gpio: this.gpioPin,
      leds: this.numPixels(),
      type: stripType,
      resetOnExit: true,
    })
    this.initialized = true
  }

  applyLights(lights: Pixel[], brightness: number): void {
    const ranges = this.getRanges()

    for (const [hwIdx, pixelIdx] of ranges) {
      if (this.skip.has(pixelIdx)) {
        this.pixelArray[hwIdx] = 0
        continue
      }
      const px = lights[pixelIdx]
      const [r, g, b] = hsvToRgb(px.hue, px.sat, brightness * px.val)
      const ri = Math.round(r * 255)
      const gi = Math.round(g * 255)
      const bi = Math.round(b * 255)

      if (this.bpp === 4) {
        const wi = Math.round(px.white * brightness * 255)
        this.pixelArray[hwIdx] = packRgbw(ri, gi, bi, wi)
      } else {
        this.pixelArray[hwIdx] = packRgb(ri, gi, bi)
      }
    }

    if (isDevMode) {
      const preview = Array.from(this.pixelArray).map(c => {
        const r = (c >> 16) & 0xff
        const g = (c >> 8) & 0xff
        const b = c & 0xff
        return termColor(r, g, b)
      }).join('')
      process.stdout.write(preview + '\r')
    } else if (ws281x) {
      ws281x.render(this.pixelArray)
    }
  }
}
